students/douglas_klos/session3/mailroom.py

```diff
@@ -0,0 +1,203 @@
+    # wouldn't deal with floats.  Try / Except was the only way I found
+    # to process possible float values.  Further I didn't want trailing
+    # or solitare 0's in the databse, such as 123.00, 123.0, 0, 0.0
+    # and needed to handle erroneous inputs.
+    while True:
+        try:
+            donation_input = float(input('Enter donation, 0 to quit: '))
+            if donation_input < 0:
+                print('Invalid input')
+            elif donation_input == 0:
+                break
+            elif donation_input.is_integer():
+                donations.append(int(donation_input))
+            else:
+                donations.append(donation_input)
+        except ValueError:
+            print('Invalid input')
+    return donations
+
+
+def add_donation():
+    """ Add new donations to a current donor """
+
+    display_database()
+    while True:
+        name = input('\nPlease select a donor to add a donation to: ')
+        if name.lower() in ('q', 'quit'):
+            return
+        elif not any(name in sub_list for sub_list in mailroom_db):
+            print('Donor not found, please select from list')
+        else:
+            break
+
+    donations = get_new_donations()
+
+    # Find the name in the database and add new donations
+    for donor, donation in mailroom_db:
+        if donor == name:
+            donation.extend(donations)
+
+
+def add_donor(name_input):
+    """ Adds name_input to the database """
+
+    # Check to see if a name was passed in, if not, read user input
+    while name_input in '':
+        name_input = input("Please enter new donor's name: ")
+        if any(name_input in sub_list for sub_list in mailroom_db):
+            print('Already exists')
+            name_input = ''
+        if name_input in ('q', 'Q'):
+            return
+    donations = get_new_donations()
+
+    mailroom_db.extend([(name_input, donations)])
+
+
+def display_database():
+    """ Displays all donors and their donations """
+
+    print()
+    for donor, donations in mailroom_db:
+        print(f'{donor:>24} : {donations:}')
+
+
+def thank_you_menu():
+    """ Menu for Thank You options """
+
+    while True:
+        name_input = input(THANK_YOU_PROMPT)
+        if name_input.lower() in ('q', 'quit'):
+            return
+        elif name_input.lower() in('l', 'list'):
+            display_database()
+        # We just want the menu again if a blank line is entered.
+        elif name_input in '':
+            pass
+        # Check to see if the name entered exists in the database
+        # If name not found, verify you want to add it.
+        elif not any(name_input in sub_list for sub_list in mailroom_db):
+            add_check = input('Donor not found, would you like to add?: ')
+            if add_check.lower() in ('y', 'yes'):
+                add_donor(name_input)
+
+        # The name was found in the database
+        else:
+            send_thank_you(name_input)
```

Comment:
I think this function and the ones it calls make this a little to complex.  Perhaps try something like this:
```
def thank_you_menu():
    name_input = input(THANK_YOU_PROMPT):

    if name_input.lower() in ('q', 'quit'):
         return
    elif name_input.lower() in('l', 'list'):
         display_database()

    for donor in mailroom_db:
        if name_input == donor[0]:
            send_thank_you(name_input)
            return

    donation_amount = float(input("Enter donation amount: "))

    mailroom_db.append((name_input, [donation_amount]))
```
I believe the above should work and you can remove a few functions, and then within `send_thank_you` you can append a donation to the end of a donor using the same `for` loop logic i have above.

Reply:
I did clean up my code a bit in that area, getting rid of some of the function calls, and cleaning out 20+ lines of code in the process.  I wanted to maintain the error trapping though to check for invalid inputs from the user.